src/modules/csm/csm.py

```diff
@@ -0,0 +1,308 @@
+        # when the tree has only one leaf.
+        stone = NodeOperatorId(self.w3.csm.module.MAX_OPERATORS_COUNT)
+        if shares:
+            shares[stone] = 0
+        if stone in shares and len(shares) > 2:
+            shares.pop(stone)
+
+        if distributed:
+            tree = Tree.new(tuple((no_id, amount) for (no_id, amount) in shares.items()))
+            logger.info({"msg": "New tree built for the report", "root": repr(tree.root)})
+            cid = self.w3.ipfs.publish(tree.encode())
+            root = tree.root
+
+        if root == ZERO_HASH:
+            logger.info({"msg": "No fee distributed so far, and tree doesn't exist"})
+
+        return ReportData(
+            self.CONSENSUS_VERSION,
+            blockstamp.ref_slot,
+            tree_root=root,
+            tree_cid=cid,
+            distributed=distributed,
+        ).as_tuple()
+
+    def is_main_data_submitted(self, blockstamp: BlockStamp) -> bool:
+        last_ref_slot = self.w3.csm.get_csm_last_processing_ref_slot(blockstamp)
+        ref_slot = self.get_current_frame(blockstamp).ref_slot
+        return last_ref_slot == ref_slot
+
+    def is_contract_reportable(self, blockstamp: BlockStamp) -> bool:
+        return not self.is_main_data_submitted(blockstamp) and not self.w3.csm.module.is_paused()
+
+    def is_reporting_allowed(self, blockstamp: ReferenceBlockStamp) -> bool:
+        on_pause = self.report_contract.is_paused(blockstamp.block_hash)
+        CONTRACT_ON_PAUSE.labels("csm").set(on_pause)
+        return not on_pause
+
+    @cached_property
+    def module(self) -> StakingModule:
+        modules: list[StakingModule] = self.w3.lido_contracts.staking_router.get_staking_modules(
+            self._receive_last_finalized_slot().block_hash
+        )
+
+        for mod in modules:
+            if mod.staking_module_address == self.w3.csm.module.address:
+                return mod
+
+        raise ValueError("No CSM module found. Wrong address?")
+
+    @lru_cache(maxsize=1)
+    def module_validators_by_node_operators(self, blockstamp: BlockStamp) -> ValidatorsByNodeOperator:
+        return self.w3.lido_validators.get_module_validators_by_node_operators(
+            StakingModuleAddress(self.module.staking_module_address), blockstamp
+        )
+
+    def collect_data(self, blockstamp: BlockStamp) -> bool:
+        """Ongoing report data collection before the report ref slot and it's submission"""
+        logger.info({"msg": "Collecting data for the report"})
+
+        converter = self.converter(blockstamp)
+
+        l_ref_slot, r_ref_slot = self.current_frame_range(blockstamp)
+        logger.info({"msg": f"Frame for performance data collect: ({l_ref_slot};{r_ref_slot}]"})
+
+        # Finalized slot is the first slot of justifying epoch, so we need to take the previous
+        finalized_epoch = EpochNumber(converter.get_epoch_by_slot(blockstamp.slot_number) - 1)
```

Comment:
What if slots is missed

Reply:
it will change nothing. we do `-1` for epoch number, not for slot number